geo/src/algorithm/stitch.rs

```diff
@@ -0,0 +1,582 @@
+        write!(f, "{self:?}")
+    }
+}
+
+impl std::error::Error for LineStitchingError {}
+
+pub(crate) type PolygonStitchingResult<T> = Result<T, LineStitchingError>;
+
+// ========= Main Algo ============
+
+/// Trait to stitch together split up polygons.
+pub trait Stitch<T: GeoFloat> {
+    /// This stitching only happens along identical edges which are located in two separate
+    /// geometries.
+    ///
+    /// ┌─────x        ┌─────┐
+    /// │    /│        │     │
+    /// │   / │        │     │
+    /// │  /  │  ───►  │     │
+    /// │ /   │        │     │
+    /// │/    │        │     │
+    /// x─────┘        └─────┘
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use geo::Stitch;
+    /// use geo::{Coord, Triangle, polygon};
+    ///
+    /// let tri1 = Triangle::from([
+    ///     Coord { x: 0.0, y: 0.0 },
+    ///     Coord { x: 1.0, y: 0.0 },
+    ///     Coord { x: 0.0, y: 1.0 },
+    /// ]);
+    /// let tri2 = Triangle::from([
+    ///     Coord { x: 1.0, y: 1.0 },
+    ///     Coord { x: 1.0, y: 0.0 },
+    ///     Coord { x: 0.0, y: 1.0 },
+    /// ]);
+    ///
+    /// let result = vec![tri1, tri2].stitch_together();
+    ///
+    /// assert!(result.is_ok());
+    ///
+    /// let mp = result.unwrap();
+    ///
+    /// assert_eq!(mp.0.len(), 1);
+    ///
+    /// let poly = mp.0[0].clone();
+    /// // 4 coords + 1 duplicate for closed-ness
+    /// assert_eq!(poly.exterior().0.len(), 4 + 1);
+    ///
+    /// let expected = polygon![
+    ///     Coord { x: 0.0, y: 1.0 },
+    ///     Coord { x: 0.0, y: 0.0 },
+    ///     Coord { x: 1.0, y: 0.0 },
+    ///     Coord { x: 1.0, y: 1.0 },
+    /// ];
+    ///
+    /// assert_eq!(poly, expected);
+    /// ```
+    ///
+    /// # Additional Notes
+    ///
+    /// If you want to do something more general like a
+    /// [`Boolean Operation Union`](https://en.wikipedia.org/wiki/Boolean_operations_on_polygons)
+    /// you should use the trait `BooleanOps` or `SpadeBoolops`.
+    fn stitch_together(&self) -> PolygonStitchingResult<MultiPolygon<T>>;
+}
+
+macro_rules! impl_stitch {
+    ($type:ty => $self:ident
+     fn stitch_together(&self) -> PolygonStitchingResult<MultiPolygon<T>> {
+         $($impls:tt)*
+     }) => {
+       impl<T> Stitch<T> for &[&$type]
+       where
+           T: GeoFloat,
+       {
+           fn stitch_together(&$self) -> PolygonStitchingResult<MultiPolygon<T>> {
+                $($impls)*
+           }
+       }
+
+       impl<T> Stitch<T> for Vec<&$type>
+       where
+           T: GeoFloat,
+       {
+           fn stitch_together(&$self) -> PolygonStitchingResult<MultiPolygon<T>> {
+               $self.as_slice().stitch_together()
+           }
+       }
+
+       impl<T> Stitch<T> for &[$type]
+       where
+           T: GeoFloat,
+       {
+           fn stitch_together(&$self) -> PolygonStitchingResult<MultiPolygon<T>> {
+               $self.iter().collect::<Vec<_>>().stitch_together()
```

Comment:
Yeah I agree. I don't like macros for things like this usually.

I already experimented quiet a bit with a trait based solution like using `AsRef<[Polygon<T>]>`, but I ran into issues since I think the user might also want to use the same for `Triangle<T>`s and implementing `AsRed<[Triangle<T>]>` in addition is not possible (unfortunately).

So imo, there are two options:

- a) just expand the macros and impl the trait for all individual possibilities. That gives the user a lot to work with
- b) just impl `AsRef<[Polygon<T>]>` and leaving the needed conversions up to the user (e.g. calling `to_polygon()` on `Triangle<T>`s before using this algo)